### PR TITLE
Build dependency graph for local files

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Refactored dependency classes to clarify resolution vs loading
Implemented LocalFileMigrator.build_dependency_graph (same pattern as NotebookMigrator)

### Linked issues
https://github.com/databrickslabs/ucx/issues/1202
Resolves https://github.com/databrickslabs/ucx/issues/1360

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests
- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/source_code/files.py [hunk 1]
```diff
@@ -14,47 +19,45 @@
...
+    """The LocalFileMigrator class is responsible for fixing code files based on their language."""
 
-    def __init__(self, languages: Languages):
+    def __init__(self, languages: Languages, resolver: DependencyResolver):
         self._languages = languages
         self._extensions = {".py": Language.PYTHON, ".sql": Language.SQL}
+        self._resolver = resolver
+
+    def build_dependency_graph(self, path: Path) -> DependencyGraph:
+        dependency = self._resolver.resolve_dependency(UnresolvedDependency(str(path)))
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/files.py:54]:
```suggestion
        dependency = self._resolver.resolve_dependency(UnresolvedDependency(path))
```
use `pathlib.Path` everywhere for paths.

File: src/databricks/labs/ucx/source_code/files.py [hunk 2]
```diff
@@ -14,47 +19,45 @@
...
 
 
-class SourceFile(SourceContainer):
+class LocalFile(SourceContainer):
 
     def __init__(self, path: str, source: str, language: Language):
         self._path = path
         self._original_code = source
         # using CellLanguage so we can reuse the facilities it provides
         self._language = CellLanguage.of_language(language)
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/files.py:28]:
```suggestion
        self._language = CellLanguage.of_language(language)
        
        def is_notebook(self) -> bool:
           return 'Databricks notebook source' in self._original_code
```

we _may_ need it, at least in the stage of folder listing.

File: src/databricks/labs/ucx/source_code/files.py [hunk 3]
```diff
@@ -14,47 +19,45 @@
...
+    def build_dependency_graph(self, parent: DependencyGraph) -> None:
         if self._language is not CellLanguage.PYTHON:
             logger.warning(f"Unsupported language: {self._language.language}")
+            return
         linter = ASTLinter.parse(self._original_code)
         run_notebook_calls = PythonLinter.list_dbutils_notebook_run_calls(linter)
         notebook_paths = {PythonLinter.get_dbutils_notebook_run_path_arg(call) for call in run_notebook_calls}
         for path in notebook_paths:
-            graph.register_dependency(Dependency(ObjectType.NOTEBOOK, path))
+            parent.register_dependency(UnresolvedDependency(path))
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/files.py:38]:
```suggestion
            parent.register_notebook(path)
```

File: src/databricks/labs/ucx/source_code/files.py [hunk 4]
```diff
@@ -14,47 +19,45 @@
...
-            graph.register_dependency(Dependency(None, import_name))
-
-
-class WorkspaceFile(SourceFile):
-    pass
-
-
-class LocalFile(SourceFile):
-    pass
+            parent.register_dependency(UnresolvedDependency(import_name))
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/files.py:42]:
```suggestion
            parent.register_import(import_name)
```

File: src/databricks/labs/ucx/source_code/notebook.py [hunk 5]
```diff
@@ -163,8 +156,15 @@ def build_dependency_graph(self, parent: DependencyGraph):
...
-                parent.register_dependency(Dependency(ObjectType.NOTEBOOK, path.strip('"')))
+                path = line[start + len(command) :]
+                path = path.strip().strip("'").strip('"')
+                object_info = ObjectInfo(object_type=ObjectType.NOTEBOOK, path=path)
+                dependency = parent.resolve_object_info(object_info)
+                if dependency is not None:
+                    parent.register_dependency(dependency)
+                else:
+                    # TODO raise Advice, see https://github.com/databrickslabs/ucx/issues/1439
+                    raise ValueError(f"Invalid notebook path in %run command: {path}")
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/notebook.py:167]:
```suggestion
                parent.register_notebook(path)
```

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 6]
```diff
@@ -158,7 +204,7 @@ def path(self):
         return self._dependency.path
 
     def register_dependency(self, dependency: Dependency) -> DependencyGraph | None:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/dependencies.py:206]:
```suggestion
    def register_notebook(self, path: Path) -> DependencyGraph | None:
        container = self._resolver.resolve_notebook(self.path, path)
        ...
    
    def register_import(self, module: str) -> DependencyGraph | None:
        container = self._resolver.resolve_import(self.path, module)
        ...
```

we have to be more concrete about the kind of dependency we're registering.

**Review by nfx — CHANGES_REQUESTED** [review]
in the current state, interfaces are very brittle, leak metadata across domain boundaries, and won't scale with the plans we have. please change public interfaces of `DependencyResolver` and `DependencyGraph` as suggested in the comments.

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 7]
```diff
@@ -34,44 +26,81 @@ def __hash__(self):
...
+class NotebookDependency(ResolvedDependency, abc.ABC):
+    pass
+
+
+class WorkspaceNotebookDependency(NotebookDependency):
+    pass
+
+
+class LocalNotebookDependency(NotebookDependency):
+    pass
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/dependencies.py]: The rationale for LocalFileDependency is naming consistency with LocalFile in files.py.

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 8]
```diff
@@ -34,44 +26,81 @@ def __hash__(self):
...
 class SourceContainer(abc.ABC):
 
-    @property
     @abc.abstractmethod
-    def object_type(self) -> ObjectType:
+    def build_dependency_graph(self, parent: DependencyGraph) -> None:
         raise NotImplementedError()
 
+
+class DependencyLoader(abc.ABC):
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/dependencies.py:44]: The current implementation is defensive i.e. it makes decisions based on actual content, rather than on assumptions inferred from the call pattern. Maybe that's overkilling and we're fine failing if say a `%run` cell refers to a file instead of a notebook, or if an `import` refers to a notebook instead of a file or a package... but let's make sure that there is a genuine benefit in dropping the defensive approach ?

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 9]
```diff
@@ -106,12 +122,38 @@ def _load_source(self, object_info: ObjectInfo) -> str:
...
+    def __init__(
+        self,
+        whitelist: Whitelist,
+        local_loader: LocalLoader,
+        ws: WorkspaceClient | None,
+    ):
         self._whitelist = Whitelist() if whitelist is None else whitelist
+        self._local_loader = local_loader
+        self._workspace_loader = None if ws is None else WorkspaceLoader(ws)
         self._advices: list[Advice] = []
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/dependencies.py:114]: Sorry but:
 - we need whitelist to decide whether or not to build sub-dependencies
 - `WorkspaceClient` is needed if we stick to my proposed defensive approach

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 10]
```diff
@@ -167,18 +213,20 @@ def register_dependency(self, dependency: Dependency) -> DependencyGraph | None:
...
         self._dependencies[resolved] = child_graph
-        container = self._loader.load_dependency(resolved)
+        container = resolved.load()
         if not container:
             return None
         container.build_dependency_graph(child_graph)
         return child_graph
 
     def locate_dependency(self, dependency: Dependency) -> DependencyGraph | None:
+        return self.locate_dependency_with_path(dependency.path)
```
[the remark this answers is not in the document]
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/dependencies.py]: `locate_dependency_with_path` is used in testing

File: src/databricks/labs/ucx/source_code/files.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/files.py:26]: that would be a bit late, see my comments re loading

File: src/databricks/labs/ucx/source_code/dependencies.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nfx) [reply src/databricks/labs/ucx/source_code/dependencies.py:43]: definitely overkilling. `%run` is only for notebooks. 

image

also, you cannot `import` a notebook - it doesn't have an extension on WorkspaceFS:
image

TLDR: we need the `NotebookLoader` interface and two implementations of it.

**Review by nfx — CHANGES_REQUESTED** [review]
abstractions are still incorrect.

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 11]
```diff
@@ -1,77 +1,97 @@
...
+class DependencyLoader(abc.ABC):
+
     @abc.abstractmethod
-    def build_dependency_graph(self, graph) -> None:
+    def load_dependency(self, dependency: Dependency) -> SourceContainer | None:
         raise NotImplementedError()
 
+    @abc.abstractmethod
+    def is_notebook(self, path: Path) -> bool:
+        raise NotImplementedError()
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/dependencies.py:51]:
```suggestion
```
this method does not belong to this interface.

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 12]
```diff
@@ -1,77 +1,97 @@
...
+
+class LocalFileLoader(DependencyLoader):
+    def load_dependency(self, dependency: Dependency) -> SourceContainer | None:
+        raise NotImplementedError()
+
+    def is_file(self, path: Path) -> bool:
+        raise NotImplementedError()
+
+    def is_notebook(self, path: Path) -> bool:
+        raise NotImplementedError()
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/dependencies.py:62]:
```suggestion
```
file loader always loads files

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 13]
```diff
@@ -1,77 +1,97 @@
...
+    pass
+
+
+class WorkspaceNotebookLoader(NotebookLoader):
 
     def __init__(self, ws: WorkspaceClient):
         self._ws = ws
 
+    def is_notebook(self, path: Path):
+        object_info = self._ws.workspace.get_status(str(path))
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/dependencies.py:79]:
```suggestion
        object_info = self._ws.workspace.get_status(path.as_posix())
```
nit

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/source_code/dependencies.py (hunk 12 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/dependencies.py:63]: it's not loading a notebook, it's making sure it's not loading a notebook as a file by accident

**Review by nfx — APPROVED** [review]
let's accept the risk of `resolve_local_file` on DependencyResolver. It might be removed in further iterations.
